features/graph/rest/src/main/java/org/opennms/netmgt/graph/rest/impl/renderer/JsonGraphRenderer.java

```diff
@@ -75,40 +76,51 @@ public String render(ImmutableGraphContainer<?> graphContainer) {
-                    graph.asGenericGraph().getProperties().forEach((key, value) -> jsonGraph.put(key, value));
+    @Override
+    public String render(ImmutableGraph<?, ?> graph) {
+        final JSONObject jsonGraph = convert(graph);
+        return jsonGraph.toString();
+    }
 
-                    // Convert Edges
-                    graph.getEdges().stream().forEach(edge -> {
-                        final GenericEdge genericEdge = edge.asGenericEdge();
-                        final Map<String, Object> edgeProperties = new HashMap<>(genericEdge.getProperties());
-                        edgeProperties.put("source", genericEdge.getSource().getId());
-                        edgeProperties.put("target", genericEdge.getTarget().getId());
-                        final JSONObject jsonEdge = new JSONObject(edgeProperties);
-                        jsonEdgesArray.put(jsonEdge);
-                    });
+    public static JSONObject convert(ImmutableGraph<?, ?> graph) {
+        final JSONObject jsonGraph = new JSONObject();
+        final JSONArray jsonEdgesArray = new JSONArray();
+        final JSONArray jsonVerticesArray = new JSONArray();
+        jsonGraph.put("edges", jsonEdgesArray);
+        jsonGraph.put("vertices", jsonVerticesArray);
 
-                    // Convert Vertices
-                    graph.getVertices().stream().forEach(vertex -> {
-                        final JSONObject jsonVertex = new JSONObject(vertex.asGenericVertex().getProperties());
-                        jsonVerticesArray.put(jsonVertex);
-                    });
+        if (graph != null) {
+            graph.asGenericGraph().getProperties().forEach((key, value) -> jsonGraph.put(key, value));
 
-                    // Convert the focus
-                    final Focus defaultFocus = graph.getDefaultFocus();
-                    final JSONObject jsonFocus = new JSONObject();
-                    jsonFocus.put("type", defaultFocus.getId());
-                    jsonFocus.put("vertexIds", new JSONArray(defaultFocus.getVertexRefs()));
-                    jsonGraph.put("defaultFocus", jsonFocus);
-                }
-                jsonGraphArray.put(jsonGraph);
+            // Convert Edges
+            graph.getEdges().stream().forEach(edge -> {
+                final GenericEdge genericEdge = edge.asGenericEdge();
+                final Map<String, Object> edgeProperties = new HashMap<>(genericEdge.getProperties());
+                edgeProperties.put("source", genericEdge.getSource().getId());
+                edgeProperties.put("target", genericEdge.getTarget().getId());
+                final JSONObject jsonEdge = new JSONObject(edgeProperties);
+                jsonEdgesArray.put(jsonEdge);
             });
-        return jsonContainer.toString();
+
+            // Convert Vertices
+            graph.getVertices().stream().forEach(vertex -> {
+                final JSONObject jsonVertex = new JSONObject(vertex.asGenericVertex().getProperties());
```

Comment:
I wonder if we need here a similar converter mechanism as for serializing into the database: https://github.com/OpenNMS/opennms/blob/jira/NMS-12376-smoke/features/graph/service/src/main/java/org/opennms/netmgt/graph/persistence/converter/ConverterService.java

Reply:
Good point.
Probably yes and no.
For now we don't need it, as all properties are converted using its `.toString()`-method.
As we are only storing primitive datatypes in the map, this is not required.
However, later on we may store more complex objects, which may need to be manually serialized.

Reply:
But what about Collections? We implemented storing Collections in HZN-1523.

Reply:
This code is serializing the properties to JSON, not a persistable String. Therefore the logic is different. At the moment we only save primitive datatypes/enums (or collections of primitive datatypes/enums) which JSONObject already serializes correctly. For other objects (e.g. IpInfo) it may serialize correctly as well. Therefore at the moment there is no need to implement a more generic solution. At a later state, this may vary.